File: frontend/src/stores/incidents.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/api/client'

export interface Incident {
  id: string
  source: string
  title: string
  description: string
  queue: string
  host: string | null
  service: string | null
  received_at: string
  raw_payload: Record<string, unknown>
}

export interface TakeoverInfo {
  username: string
  expires_at: string
  ttl_seconds: number
}

export interface OncallInfo {
  usernames: string[]
  ttl_seconds: number | null  // null = permanent (no expiry)
}

export const useIncidentsStore = defineStore('incidents', () => {
  const incidents = ref<Record<string, Incident[]>>({})
  const mutes = ref<Record<string, number>>({})
  const subscribers = ref<Record<string, string[]>>({})
  const takeovers = ref<Record<string, TakeoverInfo>>({})
  const oncall = ref<Record<string, OncallInfo>>({})
  const flushAfter = ref<Record<string, number>>({})
  const nextSchedulerRun = ref<string | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function fetchIncidents(): Promise<void> {
    loading.value = true
    error.value = null
    try {
      const res = await api.get<{
        incidents: Record<string, Incident[]>
        mutes: Record<string, number>
        subscribers: Record<string, string[]>
        takeovers: Record<string, TakeoverInfo>
        oncall: Record<string, OncallInfo>
        flush_after: Record<string, number>
        next_scheduler_run: string | null
      }>('/incidents/')
      incidents.value = res.data.incidents
      mutes.value = res.data.mutes
      subscribers.value = res.data.subscribers
      takeovers.value = res.data.takeovers
      oncall.value = res.data.oncall
      flushAfter.value = res.data.flush_after
      nextSchedulerRun.value = res.data.next_scheduler_run
    } catch {
      error.value = 'Failed to load incidents'
    } finally {
      loading.value = false
    }
  }

  async function acknowledge(incidentId: string, queue: string): Promise<void> {
    await api.post('/incidents/acknowledge', { incident_id: incidentId, queue })
    if (incidents.value[queue]) {
      incidents.value[queue] = incidents.value[queue].filter((i) => i.id !== incidentId)
    }
  }

  async function flush(queue: string): Promise<void> {
    await api.post('/incidents/flush', { queue })
    incidents.value[queue] = []
  }

  async function send(queue: string): Promise<void> {
    await api.post('/incidents/send', { queue })
    incidents.value[queue] = []
  }

  async function mute(queue: string, durationMinutes: number): Promise<void> {
    await api.post('/incidents/mute', { queue, duration_minutes: durationMinutes })
    mutes.value[queue] = durationMinutes * 60
  }

  async function unmute(queue: string): Promise<void> {
    await api.post('/incidents/unmute', { queue })
    mutes.value[queue] = 0
  }

  async function takeover(queue: string, durationMinutes: number): Promise<void> {
    const res = await api.post<{ username: string; expires_at: string }>('/incidents/takeover', {
      queue,
      duration_minutes: durationMinutes,
    })
    takeovers.value[queue] = {
      username: res.data.username,
      expires_at: res.data.expires_at,
      ttl_seconds: durationMinutes * 60,
    }
  }

  async function clearTakeover(queue: string): Promise<void> {
    await api.post('/incidents/takeover/clear', { queue })
    delete takeovers.value[queue]
  }

  async function setOncall(queue: string, usernames: string[], durationMinutes: number | null): Promise<void> {
    await api.put(`/queues/${queue}/oncall`, { usernames, duration_minutes: durationMinutes })
    await fetchIncidents()
  }

  async function clearOncall(queue: string): Promise<void> {
    await api.delete(`/queues/${queue}/oncall`)
    delete oncall.value[queue]
    await fetchIncidents()
  }

  return {
    incidents, mutes, subscribers, takeovers, oncall, flushAfter, nextSchedulerRun, loading, error,
    fetchIncidents, acknowledge, flush, send, mute, unmute, takeover, clearTakeover,
    setOncall, clearOncall,
  }
})

File: backend/app/routers/incidents.py
from fastapi import APIRouter, Depends, HTTPException

from app.config import get_config
from app.dispatch import dispatch
from app.scheduler import get_next_run_time
from app.history import log_history
from app.models import AcknowledgeRequest, FlushRequest, MuteRequest, TakeoverRequest
from app.routers.auth import get_current_user
from app.store import (
    acknowledge_incident,
    clear_takeover,
    flush_queue,
    get_all_incidents,
    get_mute_ttl,
    get_oncall,
    get_oncall_ttl,
    get_takeover,
    get_takeover_ttl,
    is_muted,
    set_mute,
    set_takeover,
    unmute_queue,
)
from app.ws import manager

router = APIRouter(tags=["incidents"])

_NOTIFY = {"type": "incidents_updated"}


@router.get("/")
async def list_incidents(username: str = Depends(get_current_user)) -> dict:
    all_incidents = await get_all_incidents()
    config = get_config()

    mute_info: dict[str, int] = {}
    subscribers: dict[str, list[str]] = {}
    takeovers: dict[str, dict] = {}
    oncall_info: dict[str, dict] = {}

    flush_after: dict[str, int] = {q.name: q.flush_after_minutes for q in config.queues}

    for queue_name in all_incidents:
        mute_info[queue_name] = await get_mute_ttl(queue_name)
        oncall = await get_oncall(queue_name)
        if oncall is not None:
            subscribers[queue_name] = oncall
            ttl = await get_oncall_ttl(queue_name)
            oncall_info[queue_name] = {"usernames": oncall, "ttl_seconds": ttl}
        else:
            subscribers[queue_name] = [u.username for u in config.users if queue_name in u.queues]
        takeover = await get_takeover(queue_name)
        if takeover:
            takeovers[queue_name] = {
                "username": takeover.username,
                "expires_at": takeover.expires_at.isoformat(),
                "ttl_seconds": await get_takeover_ttl(queue_name),
            }

    return {
        "incidents": {q: [i.model_dump() for i in incidents] for q, incidents in all_incidents.items()},
        "mutes": mute_info,
        "subscribers": subscribers,
        "takeovers": takeovers,
        "oncall": oncall_info,
        "flush_after": flush_after,
        "next_scheduler_run": get_next_run_time(),
    }


@router.post("/acknowledge")
async def acknowledge(request: AcknowledgeRequest, username: str = Depends(get_current_user)) -> dict:
    incident = await acknowledge_incident(request.incident_id, request.queue)
    if not incident:
        raise HTTPException(status_code=404, detail="Incident not found")
    await log_history(
        queue=request.queue,
        action="acknowledged",
        incidents=[incident],
        notifications=[],
        takeover_user=None,
        triggered_by=username,
    )
    await manager.broadcast(_NOTIFY)
    return {"ok": True}


@router.post("/flush")
async def flush(request: FlushRequest, username: str = Depends(get_current_user)) -> dict:
    """Acknowledge all — remove all incidents from queue without sending notifications."""
    incidents = await flush_queue(request.queue)
    if incidents:
        await log_history(
            queue=request.queue,
            action="flushed",
            incidents=incidents,
            notifications=[],
            takeover_user=None,
            triggered_by=username,
        )
    await manager.broadcast(_NOTIFY)
    return {"ok": True, "flushed": len(incidents)}


@router.post("/send")
async def send(request: FlushRequest, username: str = Depends(get_current_user)) -> dict:
    """Send all incidents to notification backends, then remove from queue."""
    if await is_muted(request.queue):
        raise HTTPException(status_code=400, detail="Queue is muted — unmute first or wait for expiry")
    incidents = await flush_queue(request.queue)
    await dispatch(incidents, request.queue, action="sent", triggered_by=username)
    await manager.broadcast(_NOTIFY)
    return {"ok": True, "sent": len(incidents)}


@router.post("/mute")
async def mute(request: MuteRequest, username: str = Depends(get_current_user)) -> dict:
    await set_mute(request.queue, request.duration_minutes)
    await manager.broadcast(_NOTIFY)
    return {"ok": True, "muted_minutes": request.duration_minutes}


@router.post("/unmute")
async def unmute(request: FlushRequest, username: str = Depends(get_current_user)) -> dict:
    await unmute_queue(request.queue)
    await manager.broadcast(_NOTIFY)
    return {"ok": True}


@router.post("/takeover")
async def takeover(request: TakeoverRequest, username: str = Depends(get_current_user)) -> dict:
    """Temporarily take over a queue — notifications go only to the calling user."""
    result = await set_takeover(request.queue, username, request.duration_minutes)
    await manager.broadcast(_NOTIFY)
    return {"ok": True, "username": result.username, "expires_at": result.expires_at.isoformat()}


@router.post("/takeover/clear")
async def takeover_clear(request: FlushRequest, username: str = Depends(get_current_user)) -> dict:
    await clear_takeover(request.queue)
    await manager.broadcast(_NOTIFY)
    return {"ok": True}
